### PR TITLE
CU-j77k6e Refined Types prototype

### PR DESCRIPTION
**Pull request by raulraja**
Prototype for composable type refinements built on arrow meta as a compiler plugin.
Allows compile time evaluation and runtime safe evaluation of calls to constructors.
```kotlin
value class TwitterHandle private constructor(val value: String) {
  companion object : Refined<String, TwitterHandle>(::TwitterHandle, {
    ensure(
      it.startsWith('@') to "${'$'}it should start with @",
      (it.length <= 16) to "${'$'}it should not contain more than 16 characters but it's ${'$'}{it.length} characters long",
      (("twitter" !in it) to "should not contain the word 'twitter'"),
      (("admin" !in it) to "should not contain the word 'admin'")
    )
  })
}

TwitterHandle("@admin") // compile time error `should not contain the word 'admin'`
TwitterHandle(dynamic) // compile time error `Prefer TwitterHandle.orNull(dynamic) or explicitly TwitterHandle.require(dynamic) for an exceptional API  
```
Composition may look like 
```kotlin
@JvmInline
value class PositiveInt private constructor(val value: Int) {
  companion object : Refined<Int, PositiveInt>(::PositiveInt, {
    ensure((it > 0) to "$it should be > 0")
  })
}

@JvmInline
value class Even private constructor(val value: Int) {
  companion object : Refined<Int, Even>(::Even, {
    ensure((it % 2 == 0) to "$it should be even")
  })
}

@JvmInline
value class PositiveEven private constructor(val value: Int) {
  companion object : Refined<Int, PositiveEven>(::PositiveEven, PositiveInt, Even)
}
```
Missing support for compile-time validation of composed local and remote predicates, will be addressed in a separate PR.

Includes a custom impl of the Kotlin compiler `IRInterpreter` modified to load and extend the map of compile-time supported functions to arbitrary rarity. This piece is internal and exposes a single `interpret` function to be easily changed once the compiler IrInterpreter is more stable.

### CODE REVIEW

**Comment by franciscodr** [comment]
Task linked: [CU-j77k6e Refined Types prototype](https://app.clickup.com/t/j77k6e)

**Review by rachelcarmena — APPROVED** [review]
:heavy_check_mark:

**Review by i-walker — APPROVED** [review]
Thanks @raulraja 🎉 
left a nit and question

File: compiler-plugin-core/src/main/kotlin/arrow/meta/phases/codegen/ir/interpreter/builtins/IrBuiltInsMapGenerated.kt [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+private fun KCallable<*>.invokeCallable(args: Array<out Any?>): Any? =
+  call(*args)
+
+inline fun <reified A> ops(): Map<CompileTimeFunction, Function1<Array<out Any?>, *>> =
+  A::class.members.associate { it.op() }
+
+
+
+val compilerIrInterpreterFunctions: Map<CompileTimeFunction, (Array<out Any?>) -> Any?> =
```
[the remark this answers is not in the document]
Comment (raulraja) [reply compiler-plugin-core/src/main/kotlin/arrow/meta/phases/codegen/ir/interpreter/builtins/IrBuiltInsMapGenerated.kt:34]: these match automatically to any arity by spreading the args as varargs

File: compiler-plugin-core/src/main/kotlin/arrow/meta/phases/codegen/ir/interpreter/state/ExceptionState.kt [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+import arrow.meta.phases.codegen.ir.interpreter.toState
+import org.jetbrains.kotlin.ir.declarations.IrClass
+import org.jetbrains.kotlin.ir.declarations.IrProperty
+import org.jetbrains.kotlin.ir.declarations.IrSimpleFunction
+import org.jetbrains.kotlin.ir.interpreter.getLastOverridden
+import org.jetbrains.kotlin.ir.util.isSubclassOf
+import org.jetbrains.kotlin.ir.util.nameForIrSerialization
+import kotlin.math.min
+
+internal class ExceptionState private constructor(
```
[the remark this answers is not in the document]
Comment (raulraja) [reply compiler-plugin-core/src/main/kotlin/arrow/meta/phases/codegen/ir/interpreter/state/ExceptionState.kt:18]: this is actually from the compiler IRInterpreter non-public apis that we need to get rid off if the compiler one covers our use case in the future.

**Pull request merged**